community/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/ConfiguringPageCacheFactory.java

```diff
@@ -69,10 +71,17 @@ protected PageCache createPageCache()
 
     public int calculateMaxPages( Config config )
     {
-        long availableMemory = config.get( pagecache_memory );
+        long pageCacheMemory = config.get( pagecache_memory );
+        long maxHeap = Runtime.getRuntime().maxMemory();
+        if ( pageCacheMemory / maxHeap > 100 )
+        {
+            log.warn( "The memory configuration looks unbalanced. It is generally recommended to have at least " +
+                      "10 KiB of heap memory, for every 1 MiB of page cache memory. The current configuration is " +
+                      "allocating %s bytes for the page cache, and %s bytes for the heap.", pageCacheMemory, maxHeap );
+        }
         long pageSize = config.get( mapped_memory_page_size );
-        long pageCount = availableMemory / pageSize;
-        return (int) Math.min( Integer.MAX_VALUE, pageCount );
+        long pageCount = pageCacheMemory / pageSize;
+        return (int) Math.min( Integer.MAX_VALUE - 2000, pageCount );
```

Comment:
I have learned through trial and error, that the array size limitation is something slightly less than Integer.MAX_VALUE.